pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -209,19 +228,65 @@ private void readAllExistingMessages(Reader<T> reader, CompletableFuture<Reader<
                                messagesRead,
                                durationMillis / 1000.0);
                        future.complete(reader);
-                       readTailMessages(reader);
+                       readTailMessages();
                    }
                 });
     }
 
-    private void readTailMessages(Reader<T> reader) {
+    private CompletableFuture<Reader<T>> createReaderAsync(MessageId startMessageId) {
+        return client.newReader(schema)
+                .topic(conf.getTopicName())
+                .startMessageId(startMessageId)
+                .autoUpdatePartitions(true)
+                .autoUpdatePartitionsInterval((int) conf.getAutoUpdatePartitionsSeconds(), TimeUnit.SECONDS)
+                .readCompacted(true)
+                .poolMessages(true)
+                .createAsync()
+                .thenApply((reader) -> {
+                    isActive = true;
+                    this.reader = reader;
+                    recoveryBackoff.reset();
+                    return reader;
+                });
+    }
+
+    private void recover() {
+        try {
+            Thread.sleep(recoveryBackoff.getNext());
+        } catch (InterruptedException e) {
+            log.debug("Interrupted sleep", e);
+        }
+        MessageId startMessageId = lastMessageId == null ? MessageId.earliest : lastMessageId;
+        if (reader != null) {
+            this.readerFuture = reader
+                    .closeAsync()
+                    .thenCompose((__) -> createReaderAsync(startMessageId));
+        } else {
+            this.readerFuture = createReaderAsync(startMessageId);
+        }
+        start();
+    }
+
+    private void readTailMessages() {
         reader.readNextAsync()
                 .thenAccept(msg -> {
                     handleMessage(msg);
-                    readTailMessages(reader);
+                    readTailMessages();
                 }).exceptionally(ex -> {
-                    log.info("Reader {} was interrupted", reader.getTopic());
+                    isActive = false;
+                    if (ex.getCause() instanceof PulsarClientException.AlreadyClosedException) {
+                        log.warn("Reader {} was interrupted. {}", reader.getTopic(), ex.getMessage());
+                    } else {
+                        log.error("Reader {} was interrupted.", reader.getTopic(), ex);
+                    }
+                    if (conf.isFaultTolerant()) {
+                        recover();
+                    }
                     return null;
                 });
     }
+
+    public boolean isActive() {
```

Comment:
Maybe we can use `isConnected` instead of `isActive`, just like `org.apache.pulsar.client.api.Consumer#isConnected`.